src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -274,45 +285,64 @@ InvokeHostFunctionOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
 
     // Get the entries for the footprint
     rust::Vec<CxxBuf> ledgerEntryCxxBufs;
-    UnorderedMap<LedgerKey, uint32_t> originalExpirations;
+    UnorderedMap<LedgerKey, LedgerEntryRentState> entryRentChanges;
+
     auto const& resources = mParentTx.sorobanResources();
     auto const& footprint = resources.footprint;
-    auto reserveSize = footprint.readOnly.size() + footprint.readWrite.size();
-    ledgerEntryCxxBufs.reserve(reserveSize);
+    auto footprintLength =
+        footprint.readOnly.size() + footprint.readWrite.size();
+
+    uint32_t ledgerSeq = ltx.loadHeader().current().ledgerSeq;
+    uint32_t autobumpLedgerCount =
+        sorobanConfig.stateExpirationSettings().autoBumpLedgers;
 
-    auto addReads = [&ledgerEntryCxxBufs, &ltx, &metrics, &sorobanConfig,
-                     &originalExpirations](auto const& keys) -> bool {
+    ledgerEntryCxxBufs.reserve(footprintLength);
+
+    auto addReads =
+        [&ledgerEntryCxxBufs, &ltx, &metrics, &sorobanConfig, &entryRentChanges,
+         autobumpLedgerCount](auto const& keys, bool readOnly) -> bool {
         for (auto const& lk : keys)
         {
+            size_t keySize = xdr::xdr_size(lk);
+            size_t entrySize = 0;
+            auto& entryRentChange = entryRentChanges[lk];
+            // The invariant is that the footprint is unique, so we can't
+            // accidentally override RW entry with RO flag.
+            entryRentChange.readOnly = readOnly;
             // Load without record for readOnly to avoid writing them later
             auto ltxe = ltx.loadWithoutRecord(lk, /*loadExpiredEntry=*/false);
-            size_t nByte{0};
             if (ltxe)
             {
                 auto const& le = ltxe.current();
                 auto buf = toCxxBuf(le);
-                nByte = buf.data->size();
-                // Typically invalid entry read should not happen unless some
-                // backward-incompatible change happened (e.g. reducing the
-                // contract data size limit) that renders previously valid
-                // entries no longer valid.
-                if (!validateContractLedgerEntry(le, nByte, sorobanConfig))
-                {
-                    return false;
-                }
+                entrySize = buf.data->size();
                 ledgerEntryCxxBufs.emplace_back(std::move(buf));
                 if (isSorobanEntry(le.data))
```

Comment:
I just realized that we're missing a validation check for `isSorobanExtEntry` in `TransactionFrame::validateSorobanResources`. EXTENSION_ENTRIES should not be allowed in the footprint.

Reply:
Seems like it's handled here: https://github.com/stellar/stellar-core/blob/875f47e247cefc25c8a4b3982ee65610e6a620e3/src/transactions/TransactionFrame.cpp#L619

Reply:
Argh looks like I can't read my own code...